runtime/shibuya/src/precompiles.rs

```diff
@@ -65,6 +67,24 @@ impl Contains<RuntimeCall> for WhitelistedCalls {
         }
     }
 }
+
+/// Filter that only allows whitelisted runtime call to pass through dispatch-lockdrop precompile
+pub struct WhitelistedLockdropCalls;
+
+impl Contains<RuntimeCall> for WhitelistedLockdropCalls {
+    fn contains(t: &RuntimeCall) -> bool {
+        match t {
+            RuntimeCall::Utility(pallet_utility::Call::batch { calls })
+            | RuntimeCall::Utility(pallet_utility::Call::batch_all { calls }) => calls
+                .iter()
+                .all(|call| WhitelistedLockdropCalls::contains(call)),
+            RuntimeCall::DappStaking(_) => true,
+            RuntimeCall::Assets(pallet_assets::Call::transfer { .. }) => true,
+            _ => false,
+        }
+    }
+}
```

Comment:
Two comments:

1.
Perhaps it's better to solve this like this: https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/proxy/src/lib.rs#L780

2.
`Utility` is fine, but for `DappStaking` we should only allow `unbond_and_unstake` and `withdraw_unbonded`. And are we missing `Balances` here?

Reply:
About 1.
It's also a solution but I like to have the filter as Trait so it's defined in precompile per runtime, and not directly in the precompile Trait impl

Reply:
That's ok, I'm not suggesting to change that. Trait should be kept.

Only you don't need to do this:
```Rust
        match t {
            RuntimeCall::Utility(pallet_utility::Call::batch { calls })
            | RuntimeCall::Utility(pallet_utility::Call::batch_all { calls }) => calls
                .iter()
                .all(|call| WhitelistedLockdropCalls::contains(call)),
```
It's enough to specify the two `Utility` calls, without the recursive check since the expanded origin will take care of it.

With the current approach, you rely on the trait implementation to cover the recursive cases.
With the proposed approach above, you just specify calls, and precompile implementation ensures that no forbidden call is executed even if wrapped into e.g. batch call.

Your code would look something like this:
```Rust
		let mut origin: T::RuntimeOrigin = frame_system::RawOrigin::Signed(real).into();
		origin.add_filter(move |c: &<T as frame_system::Config>::RuntimeCall| {
			let c = <T as Config>::RuntimeCall::from_ref(c);
			// We make sure the proxy call does access this pallet to change modify proxies.
                        DispatchValidator::validate_before_dispatch.... // or something like that
		});
```

Reply:
`validate_before_dispatch` ensure that the call is a [valid call and check that call is part of the filer](https://github.com/AstarNetwork/Astar/blob/90df4e5c01c581598111ee4a77e9592aeb88e5d3/primitives/src/precompiles.rs#L36). So if I remove the Filter Trait I will still need to check if call is valid.
Also it will need to add trait bound to all pallets [Call](https://github.com/paritytech/polkadot-sdk/blob/a03ef2791eadd93fbcce899997a578cd8f293b7b/bridges/bin/runtime-common/src/refund_relayer_extension.rs#L468). 
While also loosing the granularity per runtime that call filter Trait (with recursion check) offers. So I think it's an overkill solution for no added benefits

Reply:
Ok, fair enough.

I still don't like how this is done, and believe that the way pallet-utility handles it is much better.
But to be fair, xcm executor's `Transact` also doesn't handle this in the best possible way I guess.